packages/narwallets/src/lib/narwallets.ts

```diff
@@ -0,0 +1,274 @@
+    id++;
+    let promiseTimeout;
+    if (withTimeout) {
+      promiseTimeout = setTimeout(() => {
+        return reject(Error("timeout"));
+      }, 2000);
+    }
+    pendingPromises.push({
+      id,
+      code,
+      resolve,
+      reject,
+      timeout: promiseTimeout,
+    });
+
+    window.postMessage({
+      id,
+      src: "ws",
+      type: "nw",
+      code,
+      dest: "ext",
+      params,
+    });
+  });
+  return promise;
+};
+
+function findPendingPromiseById(id: number): PendingPromises | undefined {
+  return pendingPromises.filter((c) => c.id == id)[0];
+}
+
+function removePendingPromise(callback: PendingPromises) {
+  const index = pendingPromises.indexOf(callback);
+  if (index > -1) {
+    // only splice array when item is found
+    pendingPromises.splice(index, 1); // 2nd parameter means remove one item only
+  }
+}
+
+const setupNarwalletsState = (): void => {
+
+  // receive response from the extension content_script
+  window.addEventListener("message", (event) => {
+    if (event.source != window) {
+      return;
+    }
+    const { data } = event
+    // msg should be directed to the page (response from the extension, relayed from the content script)
+    if (!data || data.dest !== "page") {
+      return
+    }
+
+    if (data.id && data.type == "nw") {
+      const pendingPromise = findPendingPromiseById(data.id);
+      if (pendingPromise) {
+        removePendingPromise(pendingPromise);
+        if (pendingPromise.timeout) {
+          clearTimeout(pendingPromise.timeout);
+        }
+        if (!data.result) {
+          pendingPromise.reject("result is empty");
+        }
+        else if (data.result.err) {
+          pendingPromise.reject(data.result.err);
+        } else {
+          pendingPromise.resolve(data.result.data);
+        }
+      }
+    }
+  });
+};
+
+const Narwallets: WalletBehaviourFactory<InjectedWallet> = async ({
+  options,
+  metadata,
+  store,
+  provider,
+  emitter,
+  logger,
+}) => {
+
+  const signOut = async () => {
+    if (!(await isSignedIn())) {
+      return;
+    }
+
+    const res = await sendToNarwallets("sign-out");
+    // const res = await _state.wallet.signOut();
+    if (res === true) {
+      return;
+    }
+
+    const error = new Error(
+      typeof res.error === "string" ? res.error : res.error.type
+    );
+
+    // Prevent signing out by throwing.
+    if (error.message === "User reject") {
+      throw error;
+    }
+
+    // Continue signing out but log the issue.
+    logger.log("Failed to sign out");
+    logger.error(error);
+  };
+
+  return {
+    async signIn() {
+      const isUserSignedIn = await isSignedIn();
+      let code;
+      if (!isUserSignedIn) {
+        code = NARWALLETS_CODES.SIGN_IN;
+      } else {
+        code = NARWALLETS_CODES.GET_ACCOUNT_ID;
+      }
+      const response = await sendToNarwallets(code);
+      return [{ accountId: response }];
+    },
+
+    signOut,
```

Comment:
When the wallet gets **Locked** it seems like the user gets signed out of the dApp.

Reply:
Yes, if you check other wallets you will notice that even after the wallet is Locked it does not signOut the user automatically, but when you try to signAndSendTransaction it first prompts the user to unlock the wallet.

Reply:
I'm reviewing this. When a user locks the wallet, is not exiting the dApp, but staying on the same page. After calling signAndSendTransaction, I get the error `Failed to add message` in the case of the guestbook. Is the idea to avoid this, calling first signIn and then signAndSendTransaction or something similar?